### PR TITLE
Address issues in auto-generated CDDL specification

### PR DESCRIPTION
**Pull request by nc6**
# Description

This addresses issue https://github.com/IntersectMBO/cardano-ledger/issues/4462

- Moves some core type CDDL definitions to cardano-ledger-core
- Defines Shelley CDDL types through Huddle
- Adds huddle-based testing in Shelley
- Moves the Shelley CDDL to being auto-generated via the Huddle def
- Updates Conway to use some definitions from Shelley where they were unchanged
- Updates the gen-cddl.sh script to check Shelley definitions

One unexpected change: the huddle definition for protocol_param_updates has a size limit on the max block header size. This is consistent with the Haskell serialisers and data structure.

In addition, we add a couple of extra commits solving two issues:

- #4522 - we reintroduce the long explanations of various types which were subsequently not included in the generated CDDL.
- #4535 - we add a root element as the first element in the generated CDDL

Note that this PR now incorporates the contents of #4552 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by nc6** [comment]
Note that this is now out-of-date with  #4552 - I'll update after that one is merged.

File: eras/conway/impl/CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,15 @@
...
+## 1.16.2.1
+
+*
+
 ## 1.16.2.0
 
+* Added Test.Cardano.Ledger.Conway.CDDL with CDDL definitions in Conway.
+
+## 1.16.1.0
+
```
Comment (lehins) [inline eras/conway/impl/CHANGELOG.md]: Rebase gone wrong?

Also recent master bumps to `1.17.0.0` version:
```suggestion
## 1.17.0.0

* Added `Test.Cardano.Ledger.Conway.CDDL` with CDDL definitions in Conway.

## 1.16.1.0

```

File: eras/conway/impl/cddl-files/conway.cddl [hunk 2]
```diff
@@ -51,38 +88,43 @@ big_nint = #6.3(bounded_bytes)
...
+         auxiliary_data_set : {* transaction_index => auxiliary_data},
+         invalid_transactions : [* transaction_index]]
 
 block_no = uint .size 8
 
-bootstrap_witness = [public_key : $vkey
-                     , signature : $signature
-                     , chain_code : bytes .size 32
-                     , attributes : bytes]
+bootstrap_witness = [public_key : $vkey,
```
Comment (lehins) [inline eras/conway/impl/cddl-files/conway.cddl]: @nc6, please revert the change that switched the pretty printing in cuddle from leading to trailing.

I know you personal preference, but it does not align with the previous formatting and with our preference.

```suggestion
bootstrap_witness = [ public_key : $vkey
                    , signature : $signature
```



Other things like
```
certificate = [stake_registration
                // stake_deregistration
```
are better than
```
certificate = [stake_registration // 
               stake_deregistration // 
```

**Review by lehins — CHANGES_REQUESTED** [review]
@nc6 Please fix the formatting and I'll do the proper then. 
Sorry, I just can't look at the trailing commas they hurting my eyes.

File: eras/conway/impl/cddl-files/conway.cddl (hunk 2 above)
Comment (nc6) [reply eras/conway/impl/cddl-files/conway.cddl]: I have no personal preference here, other than for consistency. The switch to trailing commas was an attempt to better handle this scenario:

```
voter = [0, addr_keyhash
          // 1, scripthash
          // 2, addr_keyhash
          // 3, scripthash
          // 4, addr_keyhash]

voter = [0, addr_keyhash // 
         1, scripthash // 
         2, addr_keyhash // 
         3, scripthash // 
         4, addr_keyhash]
```

With the trailing commas and `//`, I could get the indices all lined up nicely. Honestly I didn't make a conscious decision to change things, I just tried a few formatting options until one looked nice enough. If you really have a strong preference for leading commas in CDDL, I can try to find something nice with that.

Comment (lehins) [reply eras/conway/impl/cddl-files/conway.cddl]: With respect to commas, I really do have a strong opinion for leading commas.

With respect to `//` we can keep it trailing, if that is easier, but if we can have this formatting instead, that would be even better:

```
voter = [  0, addr_keyhash
        // 1, scripthash
        // 2, addr_keyhash
        // 3, scripthash
        // 4, addr_keyhash
        ]
```

This would work much nicer with comments when we get them implemented (this is a pretty high priority, since users are complaining about lack thereof), eg:

```
voter = [  0, addr_keyhash ; Committee Member Hot KeyHash
        // 1, scripthash   ; Committee Member Hot ScriptHash
        // 2, addr_keyhash ; DRep KeyHash
        // 3, scripthash   ; DRep ScriptHash
        // 4, addr_keyhash ; StakePool KeyHash
        ]
```

**Comment by mpizenberg** [comment]
In addition to documentation comments, is it possible to regroup definitions please. For example, currently the conway certificates definitions are all over the place. It seems they are sorted alphabetically, which isn’t useful.

File: eras/conway/impl/cddl-files/conway.cddl (hunk 2 above)
Comment (nc6) [reply eras/conway/impl/cddl-files/conway.cddl]: Now updated with leading commas

**Comment by teodanciu** [comment]
Rebased on master, to remove the master-merge commit

**Review by lehins — APPROVED** [review]
CDDL is looking better and better :grinning:
